### PR TITLE
kubelet: fix duplicated status updates at pod cleanup

### PR DESCRIPTION
**Pull request by Random-Liu**
4th try of #21438. Based on #22117.

I finally found the root reason of the last bug. I reproduced the issue https://github.com/kubernetes/kubernetes/pull/21959#issuecomment-189383262 for several times. I find that each time a pod failed to be deleted whose containers were `exited` and `removed` before next pleg `relist`. After that, pleg didn't report any event, thus pod cache was never updated.

After some debugging, I find that:
- After each `relist`, the current pod record will be updated no matter the container state changed or not ([here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pleg/generic.go#L169));
- However, only pod record of which container state is changed will be updated ([here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pleg/generic.go#L205)).
- For those pods whose containers are not changed, their `old` and `current` pod record will be left in the same state.
- If now all containers of a pod are removed before `relist`, because there is no corresponding pod returned by [`GetPods`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pleg/generic.go#L162), the current pod record will not be updated, it is still the same with old pod record.
- Then `generateEvent` will not generate any event because the old and current states are equal ([here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pleg/generic.go#L118)).

In the 2nd commit of this PR. I changed the `setCurrent` function to let it update all current pod records at once. Before updating, it will clear all current pod records because we'll never use them any more.

This time I have run 40 nodes kubemark test overnight (about 10 hours) without any failure. (Without this fix, it usually failed after 20 or 30 minutes)

In summary:
- The 1st commit is fetched from #22117, it ensures that all pods with container state changed whose pod records will be updated.
- The 2nd commit ensures that `pleg` can detect container being removed.
- The 3rd commit fix the bug in fake docker client which leads to error `ListContainer` result.
- The 4th commit is the original PR #21438.

This time, I think this PR will not fail kubemark anymore......:)
@yujuhong

### CODE REVIEW

**Comment by yujuhong** [comment]
Thanks a lot for debugging this!
